File: sdk/rust/cyt-indexer/src/documents.rs
use crate::json_util::value_to_string;
use serde_json::Value;
use std::fmt::Write;

fn extract_level_info_value(data: &Value, results: &mut Vec<String>) {
    match data {
        Value::Object(map) => {
            let desc = map.get("description").and_then(|v| v.as_str());
            let default_val = map.get("default");
            let enums = map.get("enum").and_then(|v| v.as_array());

            if let Some(desc) = desc {
                let mut line = desc.to_string();
                if let Some(d) = default_val
                    && !d.is_null()
                {
                    let _ = write!(line, "; Default: {}", value_to_string(d));
                }
                if let Some(items) = enums
                    && !items.is_empty()
                {
                    let enums_str: Vec<String> = items.iter().map(value_to_string).collect();
                    let _ = write!(line, "; Options: {}", enums_str.join(", "));
                }
                results.push(line);
            }

            for val in map.values() {
                extract_level_info_value(val, results);
            }
        }
        Value::Array(items) => {
            for item in items {
                extract_level_info_value(item, results);
            }
        }
        _ => {}
    }
}

#[must_use]
pub fn extract_level_info(data: &Value) -> Vec<String> {
    let mut results = Vec::new();
    extract_level_info_value(data, &mut results);
    results
}

#[must_use]
pub fn extract_document_text(item_content: &Value) -> Option<String> {
    let level_lines = extract_level_info(item_content);
    if level_lines.is_empty() {
        return None;
    }
    Some(level_lines.join("\n"))
}

#[must_use]
pub fn extract_json_catalog_document(item: &Value) -> Option<String> {
    let obj = item.as_object()?;
    let content = obj.get("content")?;
    extract_document_text(content)
}

#[must_use]
pub fn extract_md_catalog_document(item: &Value) -> Option<String> {
    let obj = item.as_object()?;
    let content = obj.get("content")?;
    if content.is_null() {
        return None;
    }
    Some(value_to_string(content))
}

File: sdk/go/cytindexer/documents.go
package cytindexer

// ExtractLevelInfo recursively extracts description lines from a schema.
func ExtractLevelInfo(data any) []string {
	var results []string
	extractLevelInfoValue(data, &results)
	return results
}

func extractLevelInfoValue(data any, results *[]string) {
	switch d := data.(type) {
	case map[string]any:
		if desc, ok := d["description"].(string); ok && desc != "" {
			line := desc
			if defaultVal, ok := d["default"]; ok && defaultVal != nil {
				line += "; Default: " + ValueToString(defaultVal)
			}
			if enums, ok := AsArray(d["enum"]); ok && len(enums) > 0 {
				parts := make([]string, len(enums))
				for i, e := range enums {
					parts[i] = ValueToString(e)
				}
				line += "; Options: " + stringsJoin(parts, ", ")
			}
			*results = append(*results, line)
		}
		for _, val := range d {
			extractLevelInfoValue(val, results)
		}
	case []any:
		for _, item := range d {
			extractLevelInfoValue(item, results)
		}
	}
}

func stringsJoin(parts []string, sep string) string {
	if len(parts) == 0 {
		return ""
	}
	out := parts[0]
	for i := 1; i < len(parts); i++ {
		out += sep + parts[i]
	}
	return out
}

// ExtractDocumentText joins level info lines into document text.
func ExtractDocumentText(itemContent any) (string, bool) {
	lines := ExtractLevelInfo(itemContent)
	if len(lines) == 0 {
		return "", false
	}
	return stringsJoin(lines, "\n"), true
}

// ExtractJSONCatalogDocument extracts text from a json catalog entry.
func ExtractJSONCatalogDocument(item any) (string, bool) {
	obj, ok := AsObject(item)
	if !ok {
		return "", false
	}
	content, ok := obj["content"]
	if !ok {
		return "", false
	}
	return ExtractDocumentText(content)
}

// ExtractMDCatalogDocument extracts text from a markdown catalog entry.
func ExtractMDCatalogDocument(item any) (string, bool) {
	obj, ok := AsObject(item)
	if !ok {
		return "", false
	}
	content := obj["content"]
	if content == nil {
		return "", false
	}
	return ValueToString(content), true
}
